posthog/async_migrations/migrations/0001_events_sample_by.py

```diff
@@ -0,0 +1,133 @@
+from constance import config
+
+from ee.clickhouse.client import sync_execute
+from ee.clickhouse.sql.events import EVENTS_TABLE, EVENTS_TABLE_SQL
+from posthog.async_migrations.definition import AsyncMigrationDefinition, AsyncMigrationOperation
+from posthog.constants import AnalyticsDBMS
+from posthog.settings import CLICKHOUSE_CLUSTER, CLICKHOUSE_DATABASE
+from posthog.version_requirement import ServiceVersionRequirement
+
+TEMPORARY_TABLE_NAME = "temp_events_0001_events_sample_by"
+
+
+class Migration(AsyncMigrationDefinition):
+
+    description = "Events table migration for compatible sample by column."
+
+    posthog_min_version = "1.30.0"
+    posthog_max_version = "1.31.0"
+
+    service_version_requirements = [
+        ServiceVersionRequirement(service="clickhouse", supported_version=">=21.6.0,<21.7.0"),
+    ]
+
+    operations = [
+        AsyncMigrationOperation(
+            database=AnalyticsDBMS.CLICKHOUSE,
+            sql=EVENTS_TABLE_SQL().replace(EVENTS_TABLE, TEMPORARY_TABLE_NAME, 1),
+            rollback=f"DROP TABLE IF EXISTS {TEMPORARY_TABLE_NAME} ON CLUSTER {CLICKHOUSE_CLUSTER}",
+            resumable=True,
+        ),
+        AsyncMigrationOperation(
+            database=AnalyticsDBMS.CLICKHOUSE,
+            sql=f"""
+            INSERT INTO {TEMPORARY_TABLE_NAME}
+            SELECT * 
+            FROM {EVENTS_TABLE}
+            WHERE timestamp < toYYYYMM(now()) - 1
+            AND timestamp >= (SELECT max(timestamp) FROM {TEMPORARY_TABLE_NAME})""",
+            rollback=f"TRUNCATE TABLE IF EXISTS {TEMPORARY_TABLE_NAME} ON CLUSTER {CLICKHOUSE_CLUSTER}",
+            resumable=True,
+        ),
+        AsyncMigrationOperation(
+            database=AnalyticsDBMS.CLICKHOUSE,
+            sql=f"DETACH TABLE {EVENTS_TABLE}_mv ON CLUSTER {CLICKHOUSE_CLUSTER}",
+            rollback=f"ATTACH TABLE {EVENTS_TABLE}_mv ON CLUSTER {CLICKHOUSE_CLUSTER}",
+            side_effect=lambda: setattr(config, "MATERIALIZED_COLUMNS_ENABLED", False),
+            side_effect_rollback=lambda: setattr(config, "MATERIALIZED_COLUMNS_ENABLED", True),
+        ),
+        AsyncMigrationOperation(
+            database=AnalyticsDBMS.CLICKHOUSE,
+            sql=f"""
+            INSERT INTO {TEMPORARY_TABLE_NAME}
+            SELECT * 
+            FROM {EVENTS_TABLE}
+            WHERE timestamp >= (SELECT max(timestamp) FROM {TEMPORARY_TABLE_NAME})""",
```

Comment:
@fuziontech I removed this where clause from your PR: `timestamp >= toYYYYMM(now()) - 1` because that would flake if we crossed over to a new month since `now` here can be a long time after `now` a few lines before.

The check on max timestamp is more than enough here.